File: scripts/telemetry_logger.py
#!/usr/bin/env python3
"""
Telemetry Logger Module (Python)

Enforces structured logging schema per TELEMETRY-ARCHITECTURE.md

Usage:
    from telemetry_logger import create_logger
    logger = create_logger(service='git-proxy', environment='production')
    logger.info('Request received', trace_id=trace_id, request_path='/clone')
"""

import json
import logging
import os
import uuid
import hashlib
from datetime import datetime
from functools import wraps
from typing import Optional, Dict, Any


def generate_uuid() -> str:
    """Generates a UUID v4 string"""
    return str(uuid.uuid4())


def hash_user_id(user_id: Optional[str]) -> Optional[str]:
    """Hashes a user ID (SHA-256) for privacy-safe correlation"""
    if not user_id:
        return None
    return hashlib.sha256(str(user_id).encode()).hexdigest()


def generate_error_fingerprint(error: Optional[Exception]) -> str:
    """Generates deterministic error fingerprint for grouping similar errors"""
    if not error:
        return ''
    
    message = str(getattr(error, 'message', str(error)))
    # Try to get stack trace
    stack = ''
    import traceback
    try:
        stack = traceback.format_exc().split('\n')[0]
    except:
        pass
    
    fingerprint = f"{message}|{stack}"[:100]
    return hashlib.sha256(fingerprint.encode()).hexdigest()[:16]


class StructuredLogHandler(logging.Handler):
    """Custom logging handler that outputs JSON in telemetry schema"""
    
    def __init__(self, config: Dict[str, str]):
        super().__init__()
        self.config = config
    
    def emit(self, record: logging.LogRecord):
        """Emit a log record as JSON"""
        try:
            log_data = {
                'timestamp': datetime.utcnow().isoformat() + 'Z',
                'level': record.levelname.lower(),
                'service': self.config.get('service', 'unknown'),
                'region': self.config.get('region', 'unknown'),
                'host': self.config.get('host', 'unknown'),
                'environment': self.config.get('environment', 'unknown'),
                'trace_id': getattr(record, 'trace_id', generate_uuid()),
                'span_id': getattr(record, 'span_id', None),
                'request_id': getattr(record, 'request_id', None),
                'request_path': getattr(record, 'request_path', None),
                'request_method': getattr(record, 'request_method', None),
                'user_id_hash': getattr(record, 'user_id_hash', None),
                'session_id': getattr(record, 'session_id', None),
                'status_code': getattr(record, 'status_code', None),
                'duration_ms': getattr(record, 'duration_ms', None),
                'message': record.getMessage(),
                'error_fingerprint': getattr(record, 'error_fingerprint', ''),
                'context': getattr(record, 'context', {})
            }
            
            # Remove None values to keep JSON clean
            log_data = {k: v for k, v in log_data.items() if v is not None}
            
            print(json.dumps(log_data))
        except Exception:
            self.handleError(record)


def create_logger(
    service: str = 'unknown',
    region: Optional[str] = None,
    environment: Optional[str] = None,
    host: Optional[str] = None
) -> logging.Logger:
    """
    Creates a logger instance bound to a service/environment
    
    Args:
        service: Service name (code-server, git-proxy, oauth2-proxy, caddy)
        region: Deployment region (defaults to REGION env var)
        environment: Environment name (defaults to ENV or FLASK_ENV)
        host: Hostname (defaults to socket.gethostname())
    
    Returns:
        Logger instance configured for structured logging
    """
    import socket
    
    region = region or os.getenv('REGION', 'unknown')
    environment = environment or os.getenv('ENV') or os.getenv('FLASK_ENV', 'development')
    host = host or socket.gethostname()
    
    config = {
        'service': service,
        'region': region,
        'environment': environment,
        'host': host
    }
    
    logger = logging.getLogger(service)
    logger.setLevel(logging.DEBUG)
    
    # Remove existing handlers
    logger.handlers = []
    
    # Add structured JSON handler
    handler = StructuredLogHandler(config)
    handler.setFormatter(logging.Formatter('%(message)s'))
    logger.addHandler(handler)
    
    return logger


def telemetry_decorator(logger: logging.Logger):
    """
    Flask/Django view decorator to automatically capture telemetry
    
    Usage:
        @app.before_request
        def inject_trace():
            g.trace_id = request.headers.get('x-trace-id', generate_uuid())
        
        @app.route('/api/endpoint')
        @telemetry_decorator(logger)
        def my_endpoint():
            return {'status': 'ok'}
    """
    def decorator(f):
        @wraps(f)
        def decorated_function(*args, **kwargs):
            import time
            import traceback
            from flask import request, g
            
            trace_id = getattr(g, 'trace_id', generate_uuid())
            request_id = getattr(g, 'request_id', generate_uuid())
            start_time = time.time()
            
            try:
                # Log request
                extra = {
                    'trace_id': trace_id,
                    'request_id': request_id,
                    'request_method': request.method,
                    'request_path': request.path
                }
                logger.info('Request started', extra=extra)
                
                # Execute handler
                result = f(*args, **kwargs)
                
                # Log response
                duration_ms = int((time.time() - start_time) * 1000)
                extra['duration_ms'] = duration_ms
                extra['status_code'] = 200
                logger.info('Request completed', extra=extra)
                
                return result
            except Exception as e:
                duration_ms = int((time.time() - start_time) * 1000)
                error_fp = generate_error_fingerprint(e)
                extra = {
                    'trace_id': trace_id,
                    'request_id': request_id,
                    'request_method': request.method,
                    'request_path': request.path,
                    'duration_ms': duration_ms,
                    'error_fingerprint': error_fp,
                    'status_code': 500
                }
                logger.error(f'Request failed: {str(e)}', extra=extra)
                raise
        
        return decorated_function
    return decorator


# Example usage for standalone scripts
if __name__ == '__main__':
    logger = create_logger(
        service='example',
        environment='development'
    )
    
    trace_id = generate_uuid()
    
    # Log examples
    logger.info('Service started', extra={
        'trace_id': trace_id,
        'context': {'port': 8000}
    })
    
    logger.debug('Debug info', extra={
        'trace_id': trace_id,
        'context': {'debug_key': 'debug_value'}
    })
    
    try:
        raise ValueError('Example error')
    except Exception as e:
        logger.error('Operation failed', extra={
            'trace_id': trace_id,
            'error_fingerprint': generate_error_fingerprint(e)
        })

File: scripts/telemetry-logger.js
#!/usr/bin/env node
/**
 * Telemetry Logger Module
 * 
 * Enforces structured logging schema per TELEMETRY-ARCHITECTURE.md
 * Usage:
 *   const { createLogger } = require('./telemetry-logger');
 *   const logger = createLogger({ service: 'code-server', environment: 'production' });
 *   logger.info('Event', { trace_id, request_path, duration_ms });
 */

const fs = require('fs');
const crypto = require('crypto');

/**
 * Generates a UUID v4 string
 */
function generateUUID() {
  return crypto.randomUUID();
}

/**
 * Hashes a user ID (SHA-256) to enable privacy-safe correlation
 */
function hashUserId(userId) {
  if (!userId) return null;
  return crypto.createHash('sha256').update(String(userId)).digest('hex');
}

/**
 * Generates deterministic error fingerprint for grouping similar errors
 */
function generateErrorFingerprint(error) {
  if (!error) return '';
  const message = String(error.message || error);
  const stack = error.stack ? error.stack.split('\n')[0] : '';
  const fingerprint = `${message}|${stack}`.substring(0, 100);
  return crypto.createHash('sha256').update(fingerprint).digest('hex').substring(0, 16);
}

/**
 * Validates log object against telemetry schema
 * Throws if required fields missing or invalid
 */
function validateLogSchema(level, message, metadata = {}) {
  const requiredFields = ['timestamp', 'level', 'service', 'environment', 'trace_id', 'message'];
  
  const log = {
    timestamp: new Date().toISOString(),
    level,
    service: metadata.service || 'unknown',
    region: metadata.region || 'unknown',
    host: metadata.host || require('os').hostname(),
    environment: metadata.environment || 'unknown',
    trace_id: metadata.trace_id || generateUUID(),
    span_id: metadata.span_id || undefined,
    request_id: metadata.request_id || undefined,
    request_path: metadata.request_path || undefined,
    request_method: metadata.request_method || undefined,
    user_id_hash: metadata.user_id_hash || undefined,
    session_id: metadata.session_id || undefined,
    status_code: metadata.status_code || undefined,
    duration_ms: metadata.duration_ms || undefined,
    message,
    error_fingerprint: metadata.error_fingerprint || '',
    context: metadata.context || {}
  };

  // Validation
  if (!log.trace_id || log.trace_id.length < 8) {
    throw new Error(`Invalid trace_id: ${log.trace_id}`);
  }

  return log;
}

/**
 * Creates a logger instance bound to a service/environment
 */
function createLogger(config = {}) {
  const {
    service = 'unknown',
    region = process.env.REGION || 'unknown',
    environment = process.env.NODE_ENV || 'development',
    host = require('os').hostname()
  } = config;

  return {
    /**
     * Log at INFO level
     */
    info(message, metadata = {}) {
      const log = validateLogSchema('info', message, {
        service, region, environment, host,
        ...metadata
      });
      console.log(JSON.stringify(log));
      return log;
    },

    /**
     * Log at DEBUG level
     */
    debug(message, metadata = {}) {
      const log = validateLogSchema('debug', message, {
        service, region, environment, host,
        ...metadata
      });
      if (environment === 'development' || metadata.debug_escalated) {
        console.log(JSON.stringify(log));
      }
      return log;
    },

    /**
     * Log at WARN level
     */
    warn(message, metadata = {}) {
      const log = validateLogSchema('warn', message, {
        service, region, environment, host,
        ...metadata
      });
      console.warn(JSON.stringify(log));
      return log;
    },

    /**
     * Log at ERROR level with error object
     */
    error(message, error = null, metadata = {}) {
      const fingerprint = generateErrorFingerprint(error);
      const log = validateLogSchema('error', message, {
        service, region, environment, host,
        error_fingerprint: fingerprint,
        ...metadata
      });
      console.error(JSON.stringify(log));
      if (error && error.stack) {
        console.error(error.stack);
      }
      return log;
    },

    /**
     * Log at FATAL level (critical error)
     */
    fatal(message, error = null, metadata = {}) {
      const fingerprint = generateErrorFingerprint(error);
      const log = validateLogSchema('fatal', message, {
        service, region, environment, host,
        error_fingerprint: fingerprint,
        ...metadata
      });
      console.error(JSON.stringify(log));
      if (error && error.stack) {
        console.error(error.stack);
      }
      process.exit(1);
    },

    /**
     * Create a child logger with default metadata
     */
    withDefaults(defaults) {
      const self = this;
      return {
        info: (msg, meta) => self.info(msg, { ...defaults, ...meta }),
        debug: (msg, meta) => self.debug(msg, { ...defaults, ...meta }),
        warn: (msg, meta) => self.warn(msg, { ...defaults, ...meta }),
        error: (msg, err, meta) => self.error(msg, err, { ...defaults, ...meta }),
        fatal: (msg, err, meta) => self.fatal(msg, err, { ...defaults, ...meta })
      };
    }
  };
}

/**
 * Express middleware to extract/propagate trace_id from headers
 */
function traceMiddleware(logger) {
  return (req, res, next) => {
    // Extract or generate trace ID
    const traceId = req.headers['x-trace-id'] || generateUUID();
    const requestId = req.headers['x-request-id'] || generateUUID();
    
    // Attach to request object
    req.trace_id = traceId;
    req.request_id = requestId;
    
    // Set response headers
    res.setHeader('x-trace-id', traceId);
    res.setHeader('x-request-id', requestId);
    
    // Log request
    const startTime = Date.now();
    logger.info('Request received', {
      trace_id: traceId,
      request_id: requestId,
      request_method: req.method,
      request_path: req.path
    });

    // Capture response and log
    const originalSend = res.send;
    res.send = function(data) {
      const duration = Date.now() - startTime;
      logger.info('Response sent', {
        trace_id: traceId,
        request_id: requestId,
        request_method: req.method,
        request_path: req.path,
        status_code: res.statusCode,
        duration_ms: duration
      });
      return originalSend.call(this, data);
    };

    next();
  };
}

module.exports = {
  createLogger,
  generateUUID,
  hashUserId,
  generateErrorFingerprint,
  validateLogSchema,
  traceMiddleware
};
